File: julia/snp_subtractor.jl
using GZip
function cd_maker(file2)
	if occursin(".gz", file2)
		GZip.open(file2) do f
			controld=Dict()
			for line in eachline(f)
				if string(line[1]) != "#"
					line_array=split(line,"\t")
					chromosome=replace(line_array[1],"chr"=>"")
					variant=line_array[4]*line_array[2]*line_array[5]
					if !haskey(controld,chromosome) 
						controld[chromosome]=[variant]
					else
						push!(controld[chromosome],variant)
					end
				end
			end
			return controld
		end
	elseif occursin(".idx", file2)
		open(file2) do f
			controld=Dict()
			for line in eachline(f)
				if string(line[1]) != "#"
					line_array=split(line,",")
					chromosome=replace(line_array[1],"chr"=>"")
					variant=line_array[4]*line_array[2]*line_array[5]
					for variant in line_array[2:length(line_array)];
						if !haskey(controld,chromosome) 
							controld[chromosome]=[variant]
						else
							push!(controld[chromosome],variant)
						end
					end

				end
			end
			return controld
		end
	else 
		open(file2) do f
			controld=Dict()
			for line in eachline(f)
				if string(line[1]) != "#"
					line_array=split(line,"\t")
					chromosome=replace(line_array[1],"chr"=>"")
					variant=line_array[4]*line_array[2]*line_array[5]
					if !haskey(controld,string(chromosome)) 
						controld[string(chromosome)]=[variant]
					else
						push!(controld[string(chromosome)],variant)
					end
				end
			end
			return controld
		end
	end
end

function vcf_subtract(file,controldict,outfile="out.vcf")
	controld=controldict
	if occursin(".gz", file)
		GZip.open(file) do f
			for line in eachline(f)
				if string(line[1]) != "#"
					line_array=split(line,"\t")
					chromosome=replace(line_array[1],"chr"=>"")
					variant=line_array[4]*line_array[2]*line_array[5]
					if !haskey(controld,chromosome) 
						open(outfile, "a") do output
							write(output, string(line,"\n"))
						end
					elseif !(variant in controld[chromosome])
						open(outfile, "a") do output
							write(output, string(line,"\n"))
						end
					end
				end
			end
		end
	else
		open(file) do f
			for line in eachline(f)
				if string(line[1]) != "#"
					line_array=split(line,"\t")
					chromosome=replace(line_array[1],"chr"=>"")
					variant=line_array[4]*line_array[2]*line_array[5]
					if !haskey(controld,chromosome) 
						open(outfile, "a") do output
							write(output, string(line,"\n"))
						end
					elseif !(variant in controld[chromosome])
						open(outfile, "a") do output
							write(output, string(line,"\n"))
						end
					end
				end
			end
		end
	end
end
using TimerOutputs
const to = TimerOutput()
println("Starting file 2")
@timeit to "file2 parse" controld=cd_maker(ARGS[2])
println("file 2 donezo! See stats below")
println(to)
println("Starting subtraction...")
@timeit to "subtract" vcf_subtract(ARGS[1],controld,ARGS[3])
println("Fucking Right! Done!")
println(to)

File: python/snp_subtractor.py
def vcf_subtract(file, file2):
	import gzip
	import io
	import time
	import sys
	import os
	cd={}
	opt=''
	linecounter=0
	oldtime=time.time()
	starttime=time.time()
	if '.gz' in file:
		with io.TextIOWrapper(io.BufferedReader(gzip.open(file))) as fh:
			for line in fh:
				if '#' not in line:
					ll=line.split('\t')
					chroosome=ll[0]
					if 'chr' in chroosome:
						chroosome=chroosome.replace('chr','')
					if chroosome not in cd:
						cd[chroosome]=[ll[1]]
					elif chroosome in cd:
						cd[chroosome].append(ll[1])
	else:
		for line in fh:
			if '#' not in line:
				ll=line.split('\t')
				chroosome=ll[0]
				if 'chr' in chroosome:
					chroosome=chroosome.replace('chr','')
				if chroosome not in cd:
					cd[chroosome]=[ll[1]]
				elif chroosome in cd:
					cd[chroosome].append(ll[1])
	newtime=time.time()
	timeelapsed = newtime-starttime
	minutes = timeelapsed / 60
	displayminutes=format(minutes, '.2f') 
	sys.stdout.write("B Dictionary done in "+str(displayminutes)+" minutes\n")
	sys.stdout.flush()
	oldtime=newtime
	if '.gz' in file2:
		with io.TextIOWrapper(io.BufferedReader(gzip.open(file2))) as fh:
			for line in fh:
				if '#' not in line:
					linecounter+=1
					if linecounter % 100000 == 0:
						newtime=time.time()
						timeelapsed = newtime-oldtime
						minutes=timeelapsed / 60
						displayminutes=format(minutes, '.2f') 
						sys.stdout.write(str(linecounter)+' SNPs processed in '+str(displayminutes)+' minutes\n')
						sys.stdout.flush()
						oldtime=newtime
					ll = line.split('\t')
					chroosome=ll[0]
					if 'chr' in chroosome:
						chroosome=chroosome.replace('chr','')
					if chroosome not in cd:
						opt+=line
					elif ll[1] not in cd[chroosome]:
						opt+=line
	else:
		with open(file2) as fh:
			for line in fh:
				if '#' not in line:
					linecounter+=1
					if linecounter % 100000 == 0:
						newtime=time.time()
						timeelapsed = newtime-oldtime
						minutes=timeelapsed / 60
						displayminutes=format(minutes, '.2f') 
						sys.stdout.write(str(linecounter)+' SNPs processed in '+str(displayminutes)+' minutes\n')
						sys.stdout.flush()
						oldtime=newtime
					ll = line.split('\t')
					if 'chr' in chroosome:
						chroosome=chroosome.replace('chr','')
					if chroosome not in cd:
						opt+=line
					elif ll[1] not in cd[chroosome]:
						opt+=line
	with open('out.vcf','w') as fh:
		fh.write(opt)
	newtime=time.time()
	timeelapsed = newtime-starttime
	minutes = timeelapsed / 60
	displayminutes=format(minutes, '.2f') 
	sys.stdout.write('total run time '+str(displayminutes)+' minutes\n')
	sys.stdout.flush()
if __name__ == '__main__':
	print('Tjis will subtract vcfs A-B, any snps in B that are in A will be removed from A and A is returned')
	file2 = input('What file is your vcf to be A? ')
	import os
	import sys
	file2=file2.strip()
	if not os.path.exists(file2):
		print('Barnacles! File not found, try again')
		sys.exit()
	file = input('What file is your vcf to be B? ')
	file=file.strip()
	if not os.path.exists(file):
		print('Barnacles! File not found, try again')
		sys.exit()
	vcf_subtract(file, file2)


